diff --git a/node_modules/@types/react-relay/index.d.ts b/node_modules/@types/react-relay/index.d.ts
index 62c6c05..d5745e3 100644
--- a/node_modules/@types/react-relay/index.d.ts
+++ b/node_modules/@types/react-relay/index.d.ts
@@ -109,16 +109,24 @@ export {
 
 export type FetchPolicy = 'store-and-network' | 'network-only';
 
-interface QueryRendererProps<TOperation extends OperationType> {
+export interface QueryRendererReadyState<TResponse> {
+    error: Error | null;
+    props: TResponse | null;
+    retry: (() => void) | null;
+}
+
+export interface QueryRendererPropsBase<TOperation extends OperationType> {
     environment: IEnvironment;
     query: GraphQLTaggedNode | null | undefined;
-    render: (renderProps: {
-        error: Error | null;
-        props: TOperation['response'] | null;
-        retry: (() => void) | null;
-    }) => React.ReactNode;
+    render: (renderProps: QueryRendererReadyState<TOperation['response']>) => React.ReactNode;
     variables: TOperation['variables'];
 }
+
+export interface QueryRendererProps<TOperation extends OperationType> extends QueryRendererPropsBase<TOperation> {
+  cacheConfig?: CacheConfig | null;
+  fetchPolicy?: FetchPolicy;
+}
+
 declare class ReactRelayQueryRenderer<TOperation extends OperationType> extends React.Component<
     {
         cacheConfig?: CacheConfig | null;
@@ -128,7 +136,7 @@ declare class ReactRelayQueryRenderer<TOperation extends OperationType> extends
 export { ReactRelayQueryRenderer as QueryRenderer };
 
 declare class ReactRelayLocalQueryRenderer<TOperation extends OperationType> extends React.Component<
-    QueryRendererProps<TOperation>
+QueryRendererPropsBase<TOperation>
 > {}
 export { ReactRelayLocalQueryRenderer as LocalQueryRenderer };
 
