patch-package
--- a/node_modules/jquery-waypoints/shortcuts/sticky-elements/waypoints-sticky.js
+++ b/node_modules/jquery-waypoints/shortcuts/sticky-elements/waypoints-sticky.js
@@ -10,7 +10,7 @@ https://github.com/imakewebthings/jquery-waypoints/blob/master/licenses.txt
 (function() {
   (function(root, factory) {
     if (typeof define === 'function' && define.amd) {
-      return define(['jquery', 'waypoints'], factory);
+      return define(['jquery'], factory);
     } else {
       return factory(root.jQuery);
     }
