patch-package
--- a/node_modules/react-flickity-component/lib/index.js
+++ b/node_modules/react-flickity-component/lib/index.js
@@ -90,13 +90,6 @@ var FlickityComponent = function (_Component) {
       if (disableImagesLoaded) setFlickityToReady();else (0, _imagesloaded2.default)(carousel, setFlickityToReady);
       if (flickityRef) flickityRef(this.flkty);
     }
-  }, {
-    key: 'renderPortal',
-    value: function renderPortal() {
-      if (!this.carousel) return null;
-      var mountNode = this.carousel.querySelector('.flickity-slider');
-      if (mountNode) return (0, _reactDom.createPortal)(this.props.children, mountNode);
-    }
   }, {
     key: 'render',
     value: function render() {
@@ -107,7 +100,7 @@ var FlickityComponent = function (_Component) {
         ref: function ref(c) {
           _this3.carousel = c;
         }
-      }, this.renderPortal());
+      }, this.props.children);
     }
   }]);
 
