patch-package
--- a/node_modules/react-relay-network-modern/lib/createRequestError.js
+++ b/node_modules/react-relay-network-modern/lib/createRequestError.js
@@ -104,8 +104,6 @@ function createRequestError(req, res) {
 
   if (!res) {
     errorReason = 'Server return empty response.';
-  } else if (!res.json) {
-    errorReason = (res.text ? res.text : "Server return empty response with Status Code: ".concat(res.status, ".")) + (res ? "\n\n".concat(res.toString()) : '');
   } else if (res.errors) {
     if (req instanceof _RelayRequest.default) {
       errorReason = formatGraphQLErrors(req, res.errors);
@@ -114,6 +112,8 @@ function createRequestError(req, res) {
     }
   } else if (!res.data) {
     errorReason = 'Server return empty response.data.\n\n' + res.toString();
+  } else if (!res.json) {
+    errorReason = (res.text ? res.text : "Server return empty response with Status Code: ".concat(res.status, ".")) + (res ? "\n\n".concat(res.toString()) : '');
   }
 
   var error = new RRNLRequestError("Relay request for `".concat(req.getID(), "` failed by the following reasons:\n\n").concat(errorReason));
