diff --git a/node_modules/relay-runtime/lib/store/RelayResponseNormalizer.js b/node_modules/relay-runtime/lib/store/RelayResponseNormalizer.js
index c108ed6..69569a7 100644
--- a/node_modules/relay-runtime/lib/store/RelayResponseNormalizer.js
+++ b/node_modules/relay-runtime/lib/store/RelayResponseNormalizer.js
@@ -186,6 +186,9 @@ var RelayResponseNormalizer = /*#__PURE__*/function () {
               // legacy behavior for abstract refinements: always normalize even
               // if the type doesn't conform, but track if the type matches or not
               // for determining whether response fields are expected to be present
+							if (!data.hasOwnProperty) {
+								data = Object.assign({},data)
+							}
               var _implementsInterface = data.hasOwnProperty(abstractKey);
 
               var parentIsUnmatchedAbstractType = this._isUnmatchedAbstractType;
