diff --git a/node_modules/relay-runtime/lib/store/RelayResponseNormalizer.js b/node_modules/relay-runtime/lib/store/RelayResponseNormalizer.js
index 13ac420..8502116 100644
--- a/node_modules/relay-runtime/lib/store/RelayResponseNormalizer.js
+++ b/node_modules/relay-runtime/lib/store/RelayResponseNormalizer.js
@@ -278,7 +278,7 @@ var RelayResponseNormalizer = /*#__PURE__*/function () {
         //
         // Otherwise, missing fields usually indicate a server or user error (
         // the latter for manually constructed payloads).
-        if (process.env.NODE_ENV !== "production") {
+        if (process.env.NODE_ENV !== "production" && !(responseKey === "id" && storageKey === "id")) {
           process.env.NODE_ENV !== "production" ? warning(this._isClientExtension || parent.kind === LINKED_FIELD && parent.concreteType == null ? true : Object.prototype.hasOwnProperty.call(data, responseKey), 'RelayResponseNormalizer: Payload did not contain a value ' + 'for field `%s: %s`. Check that you are parsing with the same ' + 'query that was used to fetch the payload.', responseKey, storageKey) : void 0;
         }
 
