commit 3a437276f3972b6f94fd34003961c25fcd8be2df
Author: Christopher Pappas <damassi.pappas@gmail.com>
Date:   Sat Jan 20 12:28:09 2018 -0800

    CapsCase folders in prep for relay

diff --git a/desktop/apps/artwork/client/index.coffee b/desktop/apps/artwork/client/index.coffee
index f3ee7d696..526e81daf 100644
--- a/desktop/apps/artwork/client/index.coffee
+++ b/desktop/apps/artwork/client/index.coffee
@@ -59,7 +59,7 @@ module.exports =
         #{require '../components/partner/query.coffee'}
         #{require('../components/auction_artworks/query.coffee').auction_artworks}
         #{require('../components/auction_artworks/query.coffee').followed_artist_ids(CurrentUser.orNull())}
-        #{require('../../../components/react/current_auctions/query.js').default}
+        #{require('../../../components/react/CurrentAuctions/query.js').default}
         #{require '../components/artist_artworks/query.coffee'}
         #{require '../components/related_artworks/query.coffee'}
       """
@@ -71,7 +71,7 @@ module.exports =
       init: compact [
           require '../components/partner/index.coffee'
           require '../components/auction_artworks/index.coffee' unless context.is_closed
-          require('../../../components/react/current_auctions/index.js').default unless context.is_closed
+          require('../../../components/react/CurrentAuctions/index.js').default unless context.is_closed
           require '../components/artist_artworks/index.coffee' if context.is_closed
           require '../components/related_artworks/index.coffee' if context.is_closed
           require '../components/related_artists/index.coffee'
diff --git a/desktop/apps/artwork/components/auction_artworks/helpers.coffee b/desktop/apps/artwork/components/auction_artworks/helpers.coffee
index f1d395f45..fe8ad436e 100644
--- a/desktop/apps/artwork/components/auction_artworks/helpers.coffee
+++ b/desktop/apps/artwork/components/auction_artworks/helpers.coffee
@@ -1,5 +1,5 @@
 masonry = require '../../../../components/artwork_masonry_4_column/index.coffee'
-upcomingLabel = require('../../../../components/react/auction_block/utils/upcoming_label').default
+upcomingLabel = require('../../../../components/react/AuctionBlock/utils/upcomingLabel').default
 { ARTWORK_DISPLAY_NUM } = require './config.coffee'
 { partition, pluck, shuffle, take } = require 'underscore'
 
diff --git a/desktop/apps/artwork/stylesheets/index.styl b/desktop/apps/artwork/stylesheets/index.styl
index 4ba7d578f..9d1f30f44 100644
--- a/desktop/apps/artwork/stylesheets/index.styl
+++ b/desktop/apps/artwork/stylesheets/index.styl
@@ -6,9 +6,9 @@
 @require '../components/artists/index'
 @require '../components/auction/index'
 @require '../components/auction_artworks/index'
-@require '../../../components/react/auction_block'
+@require '../../../components/react/AuctionBlock'
 @require '../components/banner/index'
-@require '../../../components/react/current_auctions'
+@require '../../../components/react/CurrentAuctions'
 @require '../components/deep_zoom/index'
 @require '../components/doge/index'
 @require '../components/images/index'
diff --git a/desktop/apps/auction/components/App.js b/desktop/apps/auction/components/App.js
index ff7a185b2..4a7ab28a3 100644
--- a/desktop/apps/auction/components/App.js
+++ b/desktop/apps/auction/components/App.js
@@ -2,7 +2,7 @@ import DOM from 'desktop/apps/auction/components/DOM'
 import Layout from 'desktop/apps/auction/components/Layout'
 import PropTypes from 'prop-types'
 import React, { Component } from 'react'
-import ResponsiveWindow from 'desktop/components/react/responsive_window'
+import ResponsiveWindow from 'desktop/components/react/ResponsiveWindow'
 import { Provider } from 'react-redux'
 
 export default class App extends Component {
diff --git a/desktop/apps/auction/components/artwork_browser/ArtworkBrowser.js b/desktop/apps/auction/components/ArtworkBrowser/ArtworkBrowser.js
similarity index 83%
rename from desktop/apps/auction/components/artwork_browser/ArtworkBrowser.js
rename to desktop/apps/auction/components/ArtworkBrowser/ArtworkBrowser.js
index a19f4c8ef..31546a5b8 100644
--- a/desktop/apps/auction/components/artwork_browser/ArtworkBrowser.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/ArtworkBrowser.js
@@ -1,9 +1,9 @@
-import ArtworkDisplay from './main/ArtworkDisplay'
-import HeaderDesktop from './header/HeaderDesktop'
-import HeaderMobile from './header/HeaderMobile'
+import ArtworkDisplay from './Main/ArtworkDisplay'
+import HeaderDesktop from './Header/HeaderDesktop'
+import HeaderMobile from './Header/HeaderMobile'
 import PropTypes from 'prop-types'
 import React from 'react'
-import Sidebar from './sidebar/Sidebar'
+import Sidebar from './Sidebar/Sidebar'
 import block from 'bem-cn-lite'
 import { connect } from 'react-redux'
 
diff --git a/desktop/apps/auction/components/artwork_browser/ArtworkBrowser.styl b/desktop/apps/auction/components/ArtworkBrowser/ArtworkBrowser.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/ArtworkBrowser.styl
rename to desktop/apps/auction/components/ArtworkBrowser/ArtworkBrowser.styl
diff --git a/desktop/apps/auction/components/artwork_browser/ArtworksByFollowedArtists.js b/desktop/apps/auction/components/ArtworkBrowser/ArtworksByFollowedArtists.js
similarity index 89%
rename from desktop/apps/auction/components/artwork_browser/ArtworksByFollowedArtists.js
rename to desktop/apps/auction/components/ArtworkBrowser/ArtworksByFollowedArtists.js
index 6b75aa0d7..4c2856ed5 100644
--- a/desktop/apps/auction/components/artwork_browser/ArtworksByFollowedArtists.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/ArtworksByFollowedArtists.js
@@ -1,10 +1,10 @@
 import PropTypes from 'prop-types'
-import GridArtwork from 'desktop/apps/auction/components/artwork_browser/main/artwork/GridArtwork'
-import MasonryGrid from 'desktop/components/react/masonry_grid/MasonryGrid'
+import GridArtwork from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/GridArtwork'
+import MasonryGrid from 'desktop/components/react/MasonryGrid/MasonryGrid'
 import React from 'react'
 import get from 'lodash.get'
 import { Artwork } from '@artsy/reaction-force/dist/Components/Artwork'
-import { ArtworkRail } from '../artwork_rail/ArtworkRail'
+import { ArtworkRail } from '../ArtworkRail/ArtworkRail'
 import { RelayStubProvider } from 'desktop/components/react/RelayStubProvider'
 import { connect } from 'react-redux'
 
diff --git a/desktop/apps/auction/components/artwork_browser/BuyNowSaleArtworks.js b/desktop/apps/auction/components/ArtworkBrowser/BuyNowSaleArtworks.js
similarity index 93%
rename from desktop/apps/auction/components/artwork_browser/BuyNowSaleArtworks.js
rename to desktop/apps/auction/components/ArtworkBrowser/BuyNowSaleArtworks.js
index 803b770ba..242b89ddd 100644
--- a/desktop/apps/auction/components/artwork_browser/BuyNowSaleArtworks.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/BuyNowSaleArtworks.js
@@ -1,9 +1,9 @@
 import PropTypes from 'prop-types'
-import MasonryGrid from 'desktop/components/react/masonry_grid/MasonryGrid'
+import MasonryGrid from 'desktop/components/react/MasonryGrid/MasonryGrid'
 import React from 'react'
 import get from 'lodash.get'
 import { Artwork } from '@artsy/reaction-force/dist/Components/Artwork'
-import { ArtworkRail } from '../artwork_rail/ArtworkRail'
+import { ArtworkRail } from '../ArtworkRail/ArtworkRail'
 import { RelayStubProvider } from 'desktop/components/react/RelayStubProvider'
 import { connect } from 'react-redux'
 
diff --git a/desktop/apps/auction/components/artwork_browser/header/FilterSort.js b/desktop/apps/auction/components/ArtworkBrowser/Header/FilterSort.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/header/FilterSort.js
rename to desktop/apps/auction/components/ArtworkBrowser/Header/FilterSort.js
diff --git a/desktop/apps/auction/components/artwork_browser/header/HeaderDesktop.js b/desktop/apps/auction/components/ArtworkBrowser/Header/HeaderDesktop.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/header/HeaderDesktop.js
rename to desktop/apps/auction/components/ArtworkBrowser/Header/HeaderDesktop.js
diff --git a/desktop/apps/auction/components/artwork_browser/header/HeaderDesktop.styl b/desktop/apps/auction/components/ArtworkBrowser/Header/HeaderDesktop.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/header/HeaderDesktop.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Header/HeaderDesktop.styl
diff --git a/desktop/apps/auction/components/artwork_browser/header/HeaderMobile.js b/desktop/apps/auction/components/ArtworkBrowser/Header/HeaderMobile.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/header/HeaderMobile.js
rename to desktop/apps/auction/components/ArtworkBrowser/Header/HeaderMobile.js
diff --git a/desktop/apps/auction/components/artwork_browser/header/HeaderMobile.styl b/desktop/apps/auction/components/ArtworkBrowser/Header/HeaderMobile.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/header/HeaderMobile.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Header/HeaderMobile.styl
diff --git a/desktop/apps/auction/components/artwork_browser/header/__tests__/FilterSort.test.js b/desktop/apps/auction/components/ArtworkBrowser/Header/__tests__/FilterSort.test.js
similarity index 87%
rename from desktop/apps/auction/components/artwork_browser/header/__tests__/FilterSort.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Header/__tests__/FilterSort.test.js
index d302f412f..91df92739 100644
--- a/desktop/apps/auction/components/artwork_browser/header/__tests__/FilterSort.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Header/__tests__/FilterSort.test.js
@@ -1,8 +1,8 @@
 import * as filterActions from 'desktop/apps/auction/actions/artworkBrowser'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import FilterSort from 'desktop/apps/auction/components/artwork_browser/header/FilterSort'
+import FilterSort from 'desktop/apps/auction/components/ArtworkBrowser/Header/FilterSort'
 
-describe('auction/components/artwork_browser/header/FilterSort.test', () => {
+describe('auction/components/ArtworkBrowser/Header/FilterSort.test', () => {
   describe('<FilterSort />', () => {
     it('updates the sort with the current selected sort state', () => {
       const { wrapper } = renderTestComponent({
diff --git a/desktop/apps/auction/components/artwork_browser/header/__tests__/HeaderDesktop.test.js b/desktop/apps/auction/components/ArtworkBrowser/Header/__tests__/HeaderDesktop.test.js
similarity index 87%
rename from desktop/apps/auction/components/artwork_browser/header/__tests__/HeaderDesktop.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Header/__tests__/HeaderDesktop.test.js
index 24d882c7d..e54aab621 100644
--- a/desktop/apps/auction/components/artwork_browser/header/__tests__/HeaderDesktop.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Header/__tests__/HeaderDesktop.test.js
@@ -1,8 +1,8 @@
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import FilterSort from 'desktop/apps/auction/components/artwork_browser/header/FilterSort'
-import HeaderDesktop from 'desktop/apps/auction/components/artwork_browser/header/HeaderDesktop'
+import FilterSort from 'desktop/apps/auction/components/ArtworkBrowser/Header/FilterSort'
+import HeaderDesktop from 'desktop/apps/auction/components/ArtworkBrowser/Header/HeaderDesktop'
 
-describe('auction/components/artwork_browser/header/HeaderDesktop.test', () => {
+describe('auction/components/ArtworkBrowser/Header/HeaderDesktop.test', () => {
   describe('<HeaderDesktop />', () => {
     it('renders a <FilterSort /> component', () => {
       const { wrapper } = renderTestComponent({
diff --git a/desktop/apps/auction/components/artwork_browser/header/__tests__/HeaderMobile.test.js b/desktop/apps/auction/components/ArtworkBrowser/Header/__tests__/HeaderMobile.test.js
similarity index 89%
rename from desktop/apps/auction/components/artwork_browser/header/__tests__/HeaderMobile.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Header/__tests__/HeaderMobile.test.js
index ef71e0d91..86461dd7d 100644
--- a/desktop/apps/auction/components/artwork_browser/header/__tests__/HeaderMobile.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Header/__tests__/HeaderMobile.test.js
@@ -1,9 +1,9 @@
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import HeaderMobile from 'desktop/apps/auction/components/artwork_browser/header/HeaderMobile'
+import HeaderMobile from 'desktop/apps/auction/components/ArtworkBrowser/Header/HeaderMobile'
 import sinon from 'sinon'
 import { __RewireAPI__ as RoutesRewireApi } from 'desktop/apps/auction/actions/artworkBrowser'
 
-describe('auction/components/artwork_browser/header/HeaderMobile.test', () => {
+describe('auction/components/ArtworkBrowser/Header/HeaderMobile.test', () => {
   describe('<HeaderMobile />', () => {
     afterEach(() => {
       RoutesRewireApi.__ResetDependency__('metaphysics')
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/BidStatus.js b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/BidStatus.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/BidStatus.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/BidStatus.js
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/BidStatus.styl b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/BidStatus.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/BidStatus.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/BidStatus.styl
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/GridArtwork.js b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/GridArtwork.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/GridArtwork.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/GridArtwork.js
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/GridArtwork.styl b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/GridArtwork.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/GridArtwork.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/GridArtwork.styl
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/ListArtwork.js b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/ListArtwork.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/ListArtwork.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/ListArtwork.js
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/ListArtwork.styl b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/ListArtwork.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/ListArtwork.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/ListArtwork.styl
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/MasonryArtwork.js b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/MasonryArtwork.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/MasonryArtwork.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/MasonryArtwork.js
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/MasonryArtwork.styl b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/MasonryArtwork.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/MasonryArtwork.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/MasonryArtwork.styl
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/BidStatus.test.js b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/BidStatus.test.js
similarity index 84%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/BidStatus.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/BidStatus.test.js
index 98819c1c3..fbb4cdcf8 100644
--- a/desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/BidStatus.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/BidStatus.test.js
@@ -1,9 +1,9 @@
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test } from 'desktop/apps/auction/components/artwork_browser/main/artwork/BidStatus'
+import { test } from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/BidStatus'
 
 const { BidStatus } = test
 
-describe('auction/components/artwork_browser/main/artwork/BidStatus.test', () => {
+describe('auction/components/ArtworkBrowser/Main/Artwork/BidStatus.test', () => {
   describe('<BidStatus />', () => {
     it('renders sold label if artwork is sold', () => {
       const { wrapper } = renderTestComponent({
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/GridArtwork.test.js b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/GridArtwork.test.js
similarity index 89%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/GridArtwork.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/GridArtwork.test.js
index 47c0fb334..e4b05d70d 100644
--- a/desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/GridArtwork.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/GridArtwork.test.js
@@ -1,10 +1,10 @@
 import React from 'react'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test, __RewireAPI__ } from 'desktop/apps/auction/components/artwork_browser/main/artwork/GridArtwork'
+import { test, __RewireAPI__ } from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/GridArtwork'
 
 const { GridArtwork } = test
 
-describe('auction/components/artwork_browser/main/artwork/GridArtwork.test', () => {
+describe('auction/components/ArtworkBrowser/Main/Artwork/GridArtwork.test', () => {
   describe('<GridArtwork />', () => {
     const BidStatus = () => <div />
 
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/ListArtwork.test.js b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/ListArtwork.test.js
similarity index 89%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/ListArtwork.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/ListArtwork.test.js
index 49d886fb7..046acaf9a 100644
--- a/desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/ListArtwork.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/ListArtwork.test.js
@@ -1,10 +1,10 @@
 import React from 'react'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test, __RewireAPI__ } from 'desktop/apps/auction/components/artwork_browser/main/artwork/ListArtwork'
+import { test, __RewireAPI__ } from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/ListArtwork'
 
 const { ListArtwork } = test
 
-describe('auction/components/artwork_browser/main/artwork/ListArtwork.test', () => {
+describe('auction/components/ArtworkBrowser/Main/Artwork/ListArtwork.test', () => {
   describe('<ListArtwork />', () => {
     const BidStatus = () => <div />
 
diff --git a/desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/MasonryArtwork.test.js b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/MasonryArtwork.test.js
similarity index 89%
rename from desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/MasonryArtwork.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/MasonryArtwork.test.js
index 6ca9cd566..3f7e8f018 100644
--- a/desktop/apps/auction/components/artwork_browser/main/artwork/__tests__/MasonryArtwork.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/__tests__/MasonryArtwork.test.js
@@ -1,10 +1,10 @@
 import React from 'react'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test, __RewireAPI__ } from 'desktop/apps/auction/components/artwork_browser/main/artwork/MasonryArtwork'
+import { test, __RewireAPI__ } from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/MasonryArtwork'
 
 const { MasonryArtwork } = test
 
-describe('auction/components/artwork_browser/main/artwork/ListArtwork.test', () => {
+describe('auction/components/ArtworkBrowser/Main/Artwork/ListArtwork.test', () => {
   describe('<ListArtwork />', () => {
     const BidStatus = () => <div />
 
diff --git a/desktop/apps/auction/components/artwork_browser/main/ArtworkDisplay.js b/desktop/apps/auction/components/ArtworkBrowser/Main/ArtworkDisplay.js
similarity index 88%
rename from desktop/apps/auction/components/artwork_browser/main/ArtworkDisplay.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/ArtworkDisplay.js
index 15a3bbb74..18f9875f8 100644
--- a/desktop/apps/auction/components/artwork_browser/main/ArtworkDisplay.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Main/ArtworkDisplay.js
@@ -1,10 +1,10 @@
-import InfiniteScroll from 'desktop/components/react/infinite_scroll/InfiniteScroll'
+import InfiniteScroll from 'desktop/components/react/InfiniteScroll/InfiniteScroll'
 import Jump from 'desktop/components/jump/react'
-import MasonryArtwork from 'desktop/apps/auction/components/artwork_browser/main/artwork/MasonryArtwork'
-import GridArtwork from 'desktop/apps/auction/components/artwork_browser/main/artwork/GridArtwork'
-import ListArtwork from 'desktop/apps/auction/components/artwork_browser/main/artwork/ListArtwork'
-import LoadingSpinner from 'desktop/apps/auction/components/artwork_browser/main/LoadingSpinner'
-import MasonryGrid from 'desktop/components/react/masonry_grid/MasonryGrid'
+import MasonryArtwork from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/MasonryArtwork'
+import GridArtwork from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/GridArtwork'
+import ListArtwork from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/ListArtwork'
+import LoadingSpinner from 'desktop/apps/auction/components/ArtworkBrowser/Main/LoadingSpinner'
+import MasonryGrid from 'desktop/components/react/MasonryGrid/MasonryGrid'
 import PropTypes from 'prop-types'
 import React, { Component } from 'react'
 import block from 'bem-cn-lite'
diff --git a/desktop/apps/auction/components/artwork_browser/main/ArtworkDisplay.styl b/desktop/apps/auction/components/ArtworkBrowser/Main/ArtworkDisplay.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/ArtworkDisplay.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Main/ArtworkDisplay.styl
diff --git a/desktop/apps/auction/components/artwork_browser/main/LoadingSpinner.js b/desktop/apps/auction/components/ArtworkBrowser/Main/LoadingSpinner.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/LoadingSpinner.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/LoadingSpinner.js
diff --git a/desktop/apps/auction/components/artwork_browser/main/LoadingSpinner.styl b/desktop/apps/auction/components/ArtworkBrowser/Main/LoadingSpinner.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/main/LoadingSpinner.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Main/LoadingSpinner.styl
diff --git a/desktop/apps/auction/components/artwork_browser/main/__tests__/ArtworkDisplay.test.js b/desktop/apps/auction/components/ArtworkBrowser/Main/__tests__/ArtworkDisplay.test.js
similarity index 93%
rename from desktop/apps/auction/components/artwork_browser/main/__tests__/ArtworkDisplay.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Main/__tests__/ArtworkDisplay.test.js
index 183f1b91f..3823eef3b 100644
--- a/desktop/apps/auction/components/artwork_browser/main/__tests__/ArtworkDisplay.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Main/__tests__/ArtworkDisplay.test.js
@@ -1,11 +1,11 @@
 import React from 'react'
-import LoadingSpinner from 'desktop/apps/auction/components/artwork_browser/main/LoadingSpinner'
+import LoadingSpinner from 'desktop/apps/auction/components/ArtworkBrowser/Main/LoadingSpinner'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test, __RewireAPI__ } from 'desktop/apps/auction/components/artwork_browser/main/ArtworkDisplay'
+import { test, __RewireAPI__ } from 'desktop/apps/auction/components/ArtworkBrowser/Main/ArtworkDisplay'
 
 const { ArtworkDisplay } = test
 
-describe('auction/components/artwork_browser/main/ArtworkDisplay.test', () => {
+describe('auction/components/ArtworkBrowser/Main/ArtworkDisplay.test', () => {
   describe('<ArtworkDisplay />', () => {
     const InfiniteScroll = ({ children }) => <div>{children}</div> // eslint-disable-line
     const Jump = () => <div />
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/ArtistFilter.js b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/ArtistFilter.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/sidebar/ArtistFilter.js
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/ArtistFilter.js
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/ArtistFilter.styl b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/ArtistFilter.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/sidebar/ArtistFilter.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/ArtistFilter.styl
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/BasicCheckbox.js b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/BasicCheckbox.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/sidebar/BasicCheckbox.js
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/BasicCheckbox.js
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/BasicCheckbox.styl b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/BasicCheckbox.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/sidebar/BasicCheckbox.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/BasicCheckbox.styl
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/MediumFilter.js b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/MediumFilter.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/sidebar/MediumFilter.js
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/MediumFilter.js
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/MediumFilter.styl b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/MediumFilter.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/sidebar/MediumFilter.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/MediumFilter.styl
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/RangeSlider.js b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/RangeSlider.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/sidebar/RangeSlider.js
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/RangeSlider.js
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/RangeSlider.styl b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/RangeSlider.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/sidebar/RangeSlider.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/RangeSlider.styl
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/Sidebar.js b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/Sidebar.js
similarity index 69%
rename from desktop/apps/auction/components/artwork_browser/sidebar/Sidebar.js
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/Sidebar.js
index 764afc017..4b5f03c34 100644
--- a/desktop/apps/auction/components/artwork_browser/sidebar/Sidebar.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/Sidebar.js
@@ -1,6 +1,6 @@
-import ArtistFilter from 'desktop/apps/auction/components/artwork_browser/sidebar/ArtistFilter'
-import MediumFilter from 'desktop/apps/auction/components/artwork_browser/sidebar/MediumFilter'
-import RangeSlider from 'desktop/apps/auction/components/artwork_browser/sidebar/RangeSlider'
+import ArtistFilter from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/ArtistFilter'
+import MediumFilter from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/MediumFilter'
+import RangeSlider from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/RangeSlider'
 import PropTypes from 'prop-types'
 import React from 'react'
 import block from 'bem-cn-lite'
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/__tests__/ArtistFilter.test.js b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/ArtistFilter.test.js
similarity index 93%
rename from desktop/apps/auction/components/artwork_browser/sidebar/__tests__/ArtistFilter.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/ArtistFilter.test.js
index 709ce3c46..4e492a206 100644
--- a/desktop/apps/auction/components/artwork_browser/sidebar/__tests__/ArtistFilter.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/ArtistFilter.test.js
@@ -1,10 +1,10 @@
 import React from 'react'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test, __RewireAPI__ } from 'desktop/apps/auction/components/artwork_browser/sidebar/ArtistFilter'
+import { test, __RewireAPI__ } from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/ArtistFilter'
 
 const { ArtistFilter } = test
 
-describe('auction/components/artwork_browser/sidebar/ArtistFilter.test', () => {
+describe('auction/components/ArtworkBrowser/Sidebar/ArtistFilter.test', () => {
   describe('<ArtistFilter />', () => {
     const BasicCheckbox = () => <div />
 
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/__tests__/BasicCheckbox.test.js b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/BasicCheckbox.test.js
similarity index 73%
rename from desktop/apps/auction/components/artwork_browser/sidebar/__tests__/BasicCheckbox.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/BasicCheckbox.test.js
index 6d772fc69..e5f8259ae 100644
--- a/desktop/apps/auction/components/artwork_browser/sidebar/__tests__/BasicCheckbox.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/BasicCheckbox.test.js
@@ -1,8 +1,8 @@
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import BasicCheckbox from 'desktop/apps/auction/components/artwork_browser/sidebar/BasicCheckbox'
+import BasicCheckbox from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/BasicCheckbox'
 import sinon from 'sinon'
 
-describe('auction/components/artwork_browser/sidebar/BasicCheckbox.test', () => {
+describe('auction/components/ArtworkBrowser/Sidebar/BasicCheckbox.test', () => {
   describe('<FilterSort />', () => {
     it('triggers callback onClick', () => {
       const spy = sinon.spy()
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/__tests__/MediumFilter.test.js b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/MediumFilter.test.js
similarity index 89%
rename from desktop/apps/auction/components/artwork_browser/sidebar/__tests__/MediumFilter.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/MediumFilter.test.js
index 02da93203..9ffa71010 100644
--- a/desktop/apps/auction/components/artwork_browser/sidebar/__tests__/MediumFilter.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/MediumFilter.test.js
@@ -1,10 +1,10 @@
 import React from 'react'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test, __RewireAPI__ } from 'desktop/apps/auction/components/artwork_browser/sidebar/MediumFilter'
+import { test, __RewireAPI__ } from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/MediumFilter'
 
 const { MediumFilter } = test
 
-describe('auction/components/artwork_browser/sidebar/MediumFilter.test', () => {
+describe('auction/components/ArtworkBrowser/Sidebar/MediumFilter.test', () => {
   describe('<MediumFilter />', () => {
     const BasicCheckbox = () => <div />
 
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/__tests__/Sidebar.test.js b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/Sidebar.test.js
similarity index 70%
rename from desktop/apps/auction/components/artwork_browser/sidebar/__tests__/Sidebar.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/Sidebar.test.js
index ddd0e7628..f5a7c3c1f 100644
--- a/desktop/apps/auction/components/artwork_browser/sidebar/__tests__/Sidebar.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/__tests__/Sidebar.test.js
@@ -1,12 +1,12 @@
-import ArtistFilter from 'desktop/apps/auction/components/artwork_browser/sidebar/ArtistFilter'
-import MediumFilter from 'desktop/apps/auction/components/artwork_browser/sidebar/MediumFilter'
-import RangeSlider from 'desktop/apps/auction/components/artwork_browser/sidebar/RangeSlider'
+import ArtistFilter from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/ArtistFilter'
+import MediumFilter from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/MediumFilter'
+import RangeSlider from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/RangeSlider'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test } from 'desktop/apps/auction/components/artwork_browser/sidebar/Sidebar'
+import { test } from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/Sidebar'
 
 const { Sidebar } = test
 
-describe('auction/components/artwork_browser/sidebar/Sidebar.test', () => {
+describe('auction/components/ArtworkBrowser/Sidebar/Sidebar.test', () => {
   describe('<Sidebar />', () => {
     it('does not render a range slidere if sale is closed', () => {
       const { wrapper } = renderTestComponent({
diff --git a/desktop/apps/auction/components/artwork_browser/sidebar/rc_slider.styl b/desktop/apps/auction/components/ArtworkBrowser/Sidebar/rc_slider.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/sidebar/rc_slider.styl
rename to desktop/apps/auction/components/ArtworkBrowser/Sidebar/rc_slider.styl
diff --git a/desktop/apps/auction/components/artwork_browser/__tests__/ArtworkBrowser.test.js b/desktop/apps/auction/components/ArtworkBrowser/__tests__/ArtworkBrowser.test.js
similarity index 92%
rename from desktop/apps/auction/components/artwork_browser/__tests__/ArtworkBrowser.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/__tests__/ArtworkBrowser.test.js
index f99a6f54e..261659b09 100644
--- a/desktop/apps/auction/components/artwork_browser/__tests__/ArtworkBrowser.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/__tests__/ArtworkBrowser.test.js
@@ -1,17 +1,17 @@
 import * as actions from 'desktop/apps/auction/actions/artworkBrowser'
-import ArtistFilter from 'desktop/apps/auction/components/artwork_browser/sidebar/ArtistFilter'
-import GridArtwork from 'desktop/apps/auction/components/artwork_browser/main/artwork/GridArtwork'
-import ListArtwork from 'desktop/apps/auction/components/artwork_browser/main/artwork/ListArtwork'
-import FilterSort from 'desktop/apps/auction/components/artwork_browser/header/FilterSort'
-import MediumFilter from 'desktop/apps/auction/components/artwork_browser/sidebar/MediumFilter'
-import RangeSlider from 'desktop/apps/auction/components/artwork_browser/sidebar/RangeSlider'
-import Sidebar from 'desktop/apps/auction/components/artwork_browser/sidebar/Sidebar'
+import ArtistFilter from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/ArtistFilter'
+import GridArtwork from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/GridArtwork'
+import ListArtwork from 'desktop/apps/auction/components/ArtworkBrowser/Main/Artwork/ListArtwork'
+import FilterSort from 'desktop/apps/auction/components/ArtworkBrowser/Header/FilterSort'
+import MediumFilter from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/MediumFilter'
+import RangeSlider from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/RangeSlider'
+import Sidebar from 'desktop/apps/auction/components/ArtworkBrowser/Sidebar/Sidebar'
 import auctions from 'desktop/apps/auction/reducers'
 import React from 'react'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
 import { createStore } from 'redux'
 
-describe('auction/components/artwork_browser/ArtworkBrowser.test.js', () => {
+describe('auction/components/ArtworkBrowser/ArtworkBrowser.test.js', () => {
   describe('<ArtworkDisplay />', () => {
     let saleArtwork
 
diff --git a/desktop/apps/auction/components/artwork_browser/__tests__/ArtworksByFollowedArtists.test.js b/desktop/apps/auction/components/ArtworkBrowser/__tests__/ArtworksByFollowedArtists.test.js
similarity index 86%
rename from desktop/apps/auction/components/artwork_browser/__tests__/ArtworksByFollowedArtists.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/__tests__/ArtworksByFollowedArtists.test.js
index f5abd6070..5bd9d2f51 100644
--- a/desktop/apps/auction/components/artwork_browser/__tests__/ArtworksByFollowedArtists.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/__tests__/ArtworksByFollowedArtists.test.js
@@ -1,11 +1,11 @@
 import ArtworksByFollowedArtists from '../ArtworksByFollowedArtists'
-import MasonryGrid from 'desktop/components/react/masonry_grid/MasonryGrid'
+import MasonryGrid from 'desktop/components/react/MasonryGrid/MasonryGrid'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { ArtworkRail } from '../../artwork_rail/ArtworkRail'
+import { ArtworkRail } from '../../ArtworkRail/ArtworkRail'
 import { followedArtistSaleArtworks } from '../__tests__/fixtures/followedArtistSaleArtworks'
 import { cloneDeep } from 'lodash'
 
-describe('auction/components/artwork_browser/BuyNowSaleArtworks', () => {
+describe('auction/components/ArtworkBrowser/BuyNowSaleArtworks', () => {
   const data = {
     app: {
       isMobile: false,
diff --git a/desktop/apps/auction/components/artwork_browser/__tests__/BuyNowSaleArtworks.test.js b/desktop/apps/auction/components/ArtworkBrowser/__tests__/BuyNowSaleArtworks.test.js
similarity index 85%
rename from desktop/apps/auction/components/artwork_browser/__tests__/BuyNowSaleArtworks.test.js
rename to desktop/apps/auction/components/ArtworkBrowser/__tests__/BuyNowSaleArtworks.test.js
index a2de25ec0..381431162 100644
--- a/desktop/apps/auction/components/artwork_browser/__tests__/BuyNowSaleArtworks.test.js
+++ b/desktop/apps/auction/components/ArtworkBrowser/__tests__/BuyNowSaleArtworks.test.js
@@ -1,11 +1,11 @@
 import BuyNowSaleArtworks from '../BuyNowSaleArtworks'
-import MasonryGrid from 'desktop/components/react/masonry_grid/MasonryGrid'
+import MasonryGrid from 'desktop/components/react/MasonryGrid/MasonryGrid'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { ArtworkRail } from '../../artwork_rail/ArtworkRail'
+import { ArtworkRail } from '../../ArtworkRail/ArtworkRail'
 import { promotedSaleArtworks } from '../__tests__/fixtures/promotedSaleArtworks'
 import { cloneDeep } from 'lodash'
 
-describe('auction/components/artwork_browser/BuyNowSaleArtworks', () => {
+describe('auction/components/ArtworkBrowser/BuyNowSaleArtworks', () => {
   const data = {
     app: {
       isMobile: false,
diff --git a/desktop/apps/auction/components/artwork_browser/__tests__/fixtures/followedArtistSaleArtworks.js b/desktop/apps/auction/components/ArtworkBrowser/__tests__/fixtures/followedArtistSaleArtworks.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/__tests__/fixtures/followedArtistSaleArtworks.js
rename to desktop/apps/auction/components/ArtworkBrowser/__tests__/fixtures/followedArtistSaleArtworks.js
diff --git a/desktop/apps/auction/components/artwork_browser/__tests__/fixtures/promotedSaleArtworks.js b/desktop/apps/auction/components/ArtworkBrowser/__tests__/fixtures/promotedSaleArtworks.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_browser/__tests__/fixtures/promotedSaleArtworks.js
rename to desktop/apps/auction/components/ArtworkBrowser/__tests__/fixtures/promotedSaleArtworks.js
diff --git a/desktop/apps/auction/components/artwork_rail/ArtworkRail.js b/desktop/apps/auction/components/ArtworkRail/ArtworkRail.js
similarity index 100%
rename from desktop/apps/auction/components/artwork_rail/ArtworkRail.js
rename to desktop/apps/auction/components/ArtworkRail/ArtworkRail.js
diff --git a/desktop/apps/auction/components/artwork_rail/ArtworkRail.styl b/desktop/apps/auction/components/ArtworkRail/ArtworkRail.styl
similarity index 100%
rename from desktop/apps/auction/components/artwork_rail/ArtworkRail.styl
rename to desktop/apps/auction/components/ArtworkRail/ArtworkRail.styl
diff --git a/desktop/apps/auction/components/artwork_rail/__tests__/ArtworkRail.test.js b/desktop/apps/auction/components/ArtworkRail/__tests__/ArtworkRail.test.js
similarity index 92%
rename from desktop/apps/auction/components/artwork_rail/__tests__/ArtworkRail.test.js
rename to desktop/apps/auction/components/ArtworkRail/__tests__/ArtworkRail.test.js
index 3cf752492..ede87cfaf 100644
--- a/desktop/apps/auction/components/artwork_rail/__tests__/ArtworkRail.test.js
+++ b/desktop/apps/auction/components/ArtworkRail/__tests__/ArtworkRail.test.js
@@ -1,9 +1,9 @@
 import React from 'react'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
 import { ArtworkRail } from '../ArtworkRail'
-import { promotedSaleArtworks } from '../../artwork_browser/__tests__/fixtures/promotedSaleArtworks'
+import { promotedSaleArtworks } from '../../ArtworkBrowser/__tests__/fixtures/promotedSaleArtworks'
 
-describe('auction/components/artwork_rail/ArtworkRail', () => {
+describe('auction/components/ArtworkRail/ArtworkRail', () => {
   const data = {
     app: {
       auction: {
diff --git a/desktop/apps/auction/components/Layout.js b/desktop/apps/auction/components/Layout.js
index 9d9234b33..8c0a18ca5 100644
--- a/desktop/apps/auction/components/Layout.js
+++ b/desktop/apps/auction/components/Layout.js
@@ -1,11 +1,11 @@
-import ArtworksByFollowedArtists from 'desktop/apps/auction/components/artwork_browser/ArtworksByFollowedArtists'
-import BuyNowSaleArtworks from 'desktop/apps/auction/components/artwork_browser/BuyNowSaleArtworks'
-import ArtworkBrowser from 'desktop/apps/auction/components/artwork_browser/ArtworkBrowser'
-import AuctionBlock from 'desktop/components/react/auction_block/auction_block'
-import AuctionInfoContainer from 'desktop/apps/auction/components/layout/auction_info'
-import Banner from 'desktop/apps/auction/components/layout/Banner'
-import Footer from 'desktop/apps/auction/components/layout/Footer'
-import MyActiveBids from 'desktop/apps/auction/components/layout/active_bids/MyActiveBids'
+import ArtworksByFollowedArtists from 'desktop/apps/auction/components/ArtworkBrowser/ArtworksByFollowedArtists'
+import BuyNowSaleArtworks from 'desktop/apps/auction/components/ArtworkBrowser/BuyNowSaleArtworks'
+import ArtworkBrowser from 'desktop/apps/auction/components/ArtworkBrowser/ArtworkBrowser'
+import AuctionBlock from 'desktop/components/react/AuctionBlock/AuctionBlock'
+import AuctionInfoContainer from 'desktop/apps/auction/components/Layout/AuctionInfo'
+import Banner from 'desktop/apps/auction/components/Layout/Banner'
+import Footer from 'desktop/apps/auction/components/Layout/Footer'
+import MyActiveBids from 'desktop/apps/auction/components/Layout/ActiveBids/MyActiveBids'
 import PropTypes from 'prop-types'
 import React from 'react'
 import block from 'bem-cn-lite'
diff --git a/desktop/apps/auction/components/layout/active_bids/ActiveBidItem.js b/desktop/apps/auction/components/Layout/ActiveBids/ActiveBidItem.js
similarity index 100%
rename from desktop/apps/auction/components/layout/active_bids/ActiveBidItem.js
rename to desktop/apps/auction/components/Layout/ActiveBids/ActiveBidItem.js
diff --git a/desktop/apps/auction/components/layout/active_bids/ActiveBidItem.styl b/desktop/apps/auction/components/Layout/ActiveBids/ActiveBidItem.styl
similarity index 100%
rename from desktop/apps/auction/components/layout/active_bids/ActiveBidItem.styl
rename to desktop/apps/auction/components/Layout/ActiveBids/ActiveBidItem.styl
diff --git a/desktop/apps/auction/components/layout/active_bids/MyActiveBids.js b/desktop/apps/auction/components/Layout/ActiveBids/MyActiveBids.js
similarity index 95%
rename from desktop/apps/auction/components/layout/active_bids/MyActiveBids.js
rename to desktop/apps/auction/components/Layout/ActiveBids/MyActiveBids.js
index 1b50cafc5..b27a25e51 100644
--- a/desktop/apps/auction/components/layout/active_bids/MyActiveBids.js
+++ b/desktop/apps/auction/components/Layout/ActiveBids/MyActiveBids.js
@@ -1,4 +1,4 @@
-import ActiveBidItem from 'desktop/apps/auction/components/layout/active_bids/ActiveBidItem'
+import ActiveBidItem from 'desktop/apps/auction/components/Layout/ActiveBids/ActiveBidItem'
 import BidStatus from 'desktop/components/bid_status/react'
 import MeQuery from 'desktop/apps/auction/queries/me'
 import PropTypes from 'prop-types'
diff --git a/desktop/apps/auction/components/layout/active_bids/MyActiveBids.styl b/desktop/apps/auction/components/Layout/ActiveBids/MyActiveBids.styl
similarity index 100%
rename from desktop/apps/auction/components/layout/active_bids/MyActiveBids.styl
rename to desktop/apps/auction/components/Layout/ActiveBids/MyActiveBids.styl
diff --git a/desktop/apps/auction/components/layout/active_bids/__tests__/ActiveBidItem.test.js b/desktop/apps/auction/components/Layout/ActiveBids/__tests__/ActiveBidItem.test.js
similarity index 97%
rename from desktop/apps/auction/components/layout/active_bids/__tests__/ActiveBidItem.test.js
rename to desktop/apps/auction/components/Layout/ActiveBids/__tests__/ActiveBidItem.test.js
index fb3d6a010..b5d447be9 100644
--- a/desktop/apps/auction/components/layout/active_bids/__tests__/ActiveBidItem.test.js
+++ b/desktop/apps/auction/components/Layout/ActiveBids/__tests__/ActiveBidItem.test.js
@@ -1,6 +1,6 @@
 import React from 'react'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test } from 'desktop/apps/auction/components/layout/active_bids/ActiveBidItem'
+import { test } from 'desktop/apps/auction/components/Layout/ActiveBids/ActiveBidItem'
 
 const { ActiveBidItem } = test
 
diff --git a/desktop/apps/auction/components/layout/active_bids/__tests__/MyActiveBids.test.js b/desktop/apps/auction/components/Layout/ActiveBids/__tests__/MyActiveBids.test.js
similarity index 96%
rename from desktop/apps/auction/components/layout/active_bids/__tests__/MyActiveBids.test.js
rename to desktop/apps/auction/components/Layout/ActiveBids/__tests__/MyActiveBids.test.js
index bae3775b1..64216944a 100644
--- a/desktop/apps/auction/components/layout/active_bids/__tests__/MyActiveBids.test.js
+++ b/desktop/apps/auction/components/Layout/ActiveBids/__tests__/MyActiveBids.test.js
@@ -1,11 +1,11 @@
 import Auction from 'desktop/models/auction.coffee'
 import CurrentUser from 'desktop/models/current_user.coffee'
-import MyActiveBids from 'desktop/apps/auction/components/layout/active_bids/MyActiveBids'
+import MyActiveBids from 'desktop/apps/auction/components/Layout/ActiveBids/MyActiveBids'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
 import { initialState } from 'desktop/apps/auction/reducers'
 import { fabricate } from 'antigravity'
 
-describe('apps/auction/components/layout/active_bids/MyActiveBids', () => {
+describe('apps/auction/components/Layout/ActiveBids/MyActiveBids', () => {
   let baseData
 
   before(() => {
diff --git a/desktop/apps/auction/components/layout/auction_info/AuctionInfo.styl b/desktop/apps/auction/components/Layout/AuctionInfo/AuctionInfo.styl
similarity index 100%
rename from desktop/apps/auction/components/layout/auction_info/AuctionInfo.styl
rename to desktop/apps/auction/components/Layout/AuctionInfo/AuctionInfo.styl
diff --git a/desktop/apps/auction/components/layout/auction_info/AuctionInfoDesktop.js b/desktop/apps/auction/components/Layout/AuctionInfo/AuctionInfoDesktop.js
similarity index 100%
rename from desktop/apps/auction/components/layout/auction_info/AuctionInfoDesktop.js
rename to desktop/apps/auction/components/Layout/AuctionInfo/AuctionInfoDesktop.js
diff --git a/desktop/apps/auction/components/layout/auction_info/AuctionInfoMobile.js b/desktop/apps/auction/components/Layout/AuctionInfo/AuctionInfoMobile.js
similarity index 100%
rename from desktop/apps/auction/components/layout/auction_info/AuctionInfoMobile.js
rename to desktop/apps/auction/components/Layout/AuctionInfo/AuctionInfoMobile.js
diff --git a/desktop/apps/auction/components/layout/auction_info/Registration.js b/desktop/apps/auction/components/Layout/AuctionInfo/Registration.js
similarity index 100%
rename from desktop/apps/auction/components/layout/auction_info/Registration.js
rename to desktop/apps/auction/components/Layout/AuctionInfo/Registration.js
diff --git a/desktop/apps/auction/components/layout/auction_info/Registration.styl b/desktop/apps/auction/components/Layout/AuctionInfo/Registration.styl
similarity index 100%
rename from desktop/apps/auction/components/layout/auction_info/Registration.styl
rename to desktop/apps/auction/components/Layout/AuctionInfo/Registration.styl
diff --git a/desktop/apps/auction/components/layout/auction_info/__tests__/AuctionInfoContainer.test.js b/desktop/apps/auction/components/Layout/AuctionInfo/__tests__/AuctionInfoContainer.test.js
similarity index 97%
rename from desktop/apps/auction/components/layout/auction_info/__tests__/AuctionInfoContainer.test.js
rename to desktop/apps/auction/components/Layout/AuctionInfo/__tests__/AuctionInfoContainer.test.js
index 2acf7172d..998b1883c 100644
--- a/desktop/apps/auction/components/layout/auction_info/__tests__/AuctionInfoContainer.test.js
+++ b/desktop/apps/auction/components/Layout/AuctionInfo/__tests__/AuctionInfoContainer.test.js
@@ -1,6 +1,6 @@
 import React from 'react'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test, __RewireAPI__ } from 'desktop/apps/auction/components/layout/auction_info/index'
+import { test, __RewireAPI__ } from 'desktop/apps/auction/components/Layout/AuctionInfo'
 
 const { AuctionInfoContainer } = test
 
diff --git a/desktop/apps/auction/components/layout/auction_info/__tests__/AuctionInfoDesktop.test.js b/desktop/apps/auction/components/Layout/AuctionInfo/__tests__/AuctionInfoDesktop.test.js
similarity index 95%
rename from desktop/apps/auction/components/layout/auction_info/__tests__/AuctionInfoDesktop.test.js
rename to desktop/apps/auction/components/Layout/AuctionInfo/__tests__/AuctionInfoDesktop.test.js
index 31de70364..f9f17fa5c 100644
--- a/desktop/apps/auction/components/layout/auction_info/__tests__/AuctionInfoDesktop.test.js
+++ b/desktop/apps/auction/components/Layout/AuctionInfo/__tests__/AuctionInfoDesktop.test.js
@@ -1,7 +1,7 @@
 import React from 'react'
-import Registration from 'desktop/apps/auction/components/layout/auction_info/Registration'
+import Registration from 'desktop/apps/auction/components/Layout/AuctionInfo/Registration'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test, __RewireAPI__ } from 'desktop/apps/auction/components/layout/auction_info/AuctionInfoDesktop'
+import { test, __RewireAPI__ } from 'desktop/apps/auction/components/Layout/AuctionInfo/AuctionInfoDesktop'
 
 const { AuctionInfoDesktop } = test
 
diff --git a/desktop/apps/auction/components/layout/auction_info/__tests__/AuctionInfoMobile.test.js b/desktop/apps/auction/components/Layout/AuctionInfo/__tests__/AuctionInfoMobile.test.js
similarity index 95%
rename from desktop/apps/auction/components/layout/auction_info/__tests__/AuctionInfoMobile.test.js
rename to desktop/apps/auction/components/Layout/AuctionInfo/__tests__/AuctionInfoMobile.test.js
index 8717c4aa7..bc45bc131 100644
--- a/desktop/apps/auction/components/layout/auction_info/__tests__/AuctionInfoMobile.test.js
+++ b/desktop/apps/auction/components/Layout/AuctionInfo/__tests__/AuctionInfoMobile.test.js
@@ -1,5 +1,5 @@
-import AuctionInfoMobileWrapper, { test } from 'desktop/apps/auction/components/layout/auction_info/AuctionInfoMobile'
-import Registration from 'desktop/apps/auction/components/layout/auction_info/Registration'
+import AuctionInfoMobileWrapper, { test } from 'desktop/apps/auction/components/Layout/AuctionInfo/AuctionInfoMobile'
+import Registration from 'desktop/apps/auction/components/Layout/AuctionInfo/Registration'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
 import sinon from 'sinon'
 
diff --git a/desktop/apps/auction/components/layout/auction_info/__tests__/Registration.test.js b/desktop/apps/auction/components/Layout/AuctionInfo/__tests__/Registration.test.js
similarity index 96%
rename from desktop/apps/auction/components/layout/auction_info/__tests__/Registration.test.js
rename to desktop/apps/auction/components/Layout/AuctionInfo/__tests__/Registration.test.js
index 8eff403e9..36a464c34 100644
--- a/desktop/apps/auction/components/layout/auction_info/__tests__/Registration.test.js
+++ b/desktop/apps/auction/components/Layout/AuctionInfo/__tests__/Registration.test.js
@@ -1,5 +1,5 @@
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test } from 'desktop/apps/auction/components/layout/auction_info/Registration'
+import { test } from 'desktop/apps/auction/components/Layout/AuctionInfo/Registration'
 
 const { Registration } = test
 
diff --git a/desktop/apps/auction/components/layout/auction_info/index.js b/desktop/apps/auction/components/Layout/AuctionInfo/index.js
similarity index 100%
rename from desktop/apps/auction/components/layout/auction_info/index.js
rename to desktop/apps/auction/components/Layout/AuctionInfo/index.js
diff --git a/desktop/apps/auction/components/layout/Banner.js b/desktop/apps/auction/components/Layout/Banner.js
similarity index 100%
rename from desktop/apps/auction/components/layout/Banner.js
rename to desktop/apps/auction/components/Layout/Banner.js
diff --git a/desktop/apps/auction/components/layout/Banner.styl b/desktop/apps/auction/components/Layout/Banner.styl
similarity index 100%
rename from desktop/apps/auction/components/layout/Banner.styl
rename to desktop/apps/auction/components/Layout/Banner.styl
diff --git a/desktop/apps/auction/components/layout/Footer.js b/desktop/apps/auction/components/Layout/Footer.js
similarity index 100%
rename from desktop/apps/auction/components/layout/Footer.js
rename to desktop/apps/auction/components/Layout/Footer.js
diff --git a/desktop/apps/auction/components/layout/Footer.styl b/desktop/apps/auction/components/Layout/Footer.styl
similarity index 100%
rename from desktop/apps/auction/components/layout/Footer.styl
rename to desktop/apps/auction/components/Layout/Footer.styl
diff --git a/desktop/apps/auction/components/layout/__tests__/Banner.test.js b/desktop/apps/auction/components/Layout/__tests__/Banner.test.js
similarity index 95%
rename from desktop/apps/auction/components/layout/__tests__/Banner.test.js
rename to desktop/apps/auction/components/Layout/__tests__/Banner.test.js
index fcad4f9b6..9b5efea03 100644
--- a/desktop/apps/auction/components/layout/__tests__/Banner.test.js
+++ b/desktop/apps/auction/components/Layout/__tests__/Banner.test.js
@@ -1,5 +1,5 @@
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test } from 'desktop/apps/auction/components/layout/Banner'
+import { test } from 'desktop/apps/auction/components/Layout/Banner'
 import sinon from 'sinon'
 const { Banner } = test
 
diff --git a/desktop/apps/auction/components/layout/__tests__/Footer.test.js b/desktop/apps/auction/components/Layout/__tests__/Footer.test.js
similarity index 97%
rename from desktop/apps/auction/components/layout/__tests__/Footer.test.js
rename to desktop/apps/auction/components/Layout/__tests__/Footer.test.js
index ccd950c9f..94fb3351e 100644
--- a/desktop/apps/auction/components/layout/__tests__/Footer.test.js
+++ b/desktop/apps/auction/components/Layout/__tests__/Footer.test.js
@@ -1,7 +1,7 @@
 import Articles from 'desktop/collections/articles.coffee'
 import footerItems from 'desktop/apps/auction/utils/footerItems'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
-import { test } from 'desktop/apps/auction/components/layout/Footer'
+import { test } from 'desktop/apps/auction/components/Layout/Footer'
 
 const { Footer } = test
 
diff --git a/desktop/apps/auction/components/layout/meta.jade b/desktop/apps/auction/components/Layout/meta.jade
similarity index 100%
rename from desktop/apps/auction/components/layout/meta.jade
rename to desktop/apps/auction/components/Layout/meta.jade
diff --git a/desktop/apps/auction/components/__tests__/Layout.test.js b/desktop/apps/auction/components/__tests__/Layout.test.js
index 81419c5e0..389f237cf 100644
--- a/desktop/apps/auction/components/__tests__/Layout.test.js
+++ b/desktop/apps/auction/components/__tests__/Layout.test.js
@@ -5,8 +5,8 @@ import moment from 'moment'
 import renderTestComponent from 'desktop/apps/auction/__tests__/utils/renderTestComponent'
 import Layout from 'desktop/apps/auction/components/Layout'
 import { cloneDeep } from 'lodash'
-import { followedArtistSaleArtworks } from '../artwork_browser/__tests__/fixtures/followedArtistSaleArtworks'
-import { promotedSaleArtworks } from '../artwork_browser/__tests__/fixtures/promotedSaleArtworks'
+import { followedArtistSaleArtworks } from '../ArtworkBrowser/__tests__/fixtures/followedArtistSaleArtworks'
+import { promotedSaleArtworks } from '../ArtworkBrowser/__tests__/fixtures/promotedSaleArtworks'
 
 describe('<Layout />', () => {
   beforeEach(() => {
diff --git a/desktop/apps/auction/index.js b/desktop/apps/auction/index.js
index 39e62ac2b..3e5513769 100644
--- a/desktop/apps/auction/index.js
+++ b/desktop/apps/auction/index.js
@@ -5,7 +5,7 @@ import express from 'express'
 const app = module.exports = express()
 
 app.set('view engine', 'jade')
-app.set('views', `${__dirname}/components/layout`)
+app.set('views', `${__dirname}/components/Layout`)
 
 app.get('/sale/:id', routes.index)
 app.get('/sale/:id/confirm-registration', routes.index)
diff --git a/desktop/apps/auction/queries/sale.js b/desktop/apps/auction/queries/sale.js
index 24cd9d891..023494ec7 100644
--- a/desktop/apps/auction/queries/sale.js
+++ b/desktop/apps/auction/queries/sale.js
@@ -3,7 +3,7 @@ export default function SaleQuery (id) {
     sale(id: "${id}") {
       _id
       associated_sale {
-        ${require('../../../components/react/auction_block/query.js').default}
+        ${require('../../../components/react/AuctionBlock/query.js').default}
       }
       auction_state
       cover_image {
diff --git a/desktop/apps/auction/reducers/app.js b/desktop/apps/auction/reducers/app.js
index 113399340..1a6e6c6fb 100644
--- a/desktop/apps/auction/reducers/app.js
+++ b/desktop/apps/auction/reducers/app.js
@@ -5,6 +5,7 @@ export const initialState = {
   articles: [],
   auction: {},
   footerItems: [],
+  isEcommerceSale: false,
   isLiveOpen: undefined,
   isMobile: undefined,
   liveAuctionUrl: x => x,
diff --git a/desktop/apps/auction/reducers/index.js b/desktop/apps/auction/reducers/index.js
index 76e728ffe..533176169 100644
--- a/desktop/apps/auction/reducers/index.js
+++ b/desktop/apps/auction/reducers/index.js
@@ -1,7 +1,7 @@
 import app, { initialState as appInitialState } from 'desktop/apps/auction/reducers/app'
 import filter, { initialState as filterInitialState } from 'desktop/apps/auction/reducers/artworkBrowser'
 import { combineReducers } from 'redux'
-import { composeReducers, responsiveWindowReducer } from 'desktop/components/react/responsive_window'
+import { composeReducers, responsiveWindowReducer } from 'desktop/components/react/ResponsiveWindow'
 
 export const initialState = {
   ...appInitialState,
diff --git a/desktop/apps/auction/routes.js b/desktop/apps/auction/routes.js
index 232f30961..c814cc3f6 100644
--- a/desktop/apps/auction/routes.js
+++ b/desktop/apps/auction/routes.js
@@ -16,6 +16,7 @@ import { initialState as appInitialState } from 'desktop/apps/auction/reducers/a
 import { initialState as auctionWorksInitialState } from 'desktop/apps/auction/reducers/artworkBrowser'
 import { getLiveAuctionUrl } from 'utils/domain/auctions/urls'
 import { renderLayout } from '@artsy/stitch'
+import { isObject } from 'underscore'
 
 export async function index (req, res, next) {
   const saleId = req.params.id
@@ -47,7 +48,14 @@ export async function index (req, res, next) {
 
     // TODO: Refactor out Backbone
     const auctionModel = new Auction(sale)
-    const isEcommerceSale = saleId === 'chris-ecommerce-sale'
+
+    /**
+     * An e-commerce sale is one which is created by following these steps:
+     * https://github.com/artsy/auctions/blob/master/auctions_support_faq.md#how-do-i-create-an-e-commerce-sale
+     * It enables relations between one or more conventional auctions and
+     * represents "Buy Now"-like artworks
+     */
+    const isEcommerceSale = !sale.is_auction && isObject(sale.promoted_sale)
 
     // If an e-commerce sale, remove all sort options that are Auction related
     let artworkBrowserSortOptions = auctionWorksInitialState.sortMap
@@ -73,7 +81,6 @@ export async function index (req, res, next) {
       }, appInitialState),
 
       artworkBrowser: u({
-        // FIXME: This is a temporary fix to update UI
         filterParams: {
           sale_id: saleId
         },
diff --git a/desktop/apps/auction/stylesheets/index.styl b/desktop/apps/auction/stylesheets/index.styl
index 51b30c84d..737bc9570 100644
--- a/desktop/apps/auction/stylesheets/index.styl
+++ b/desktop/apps/auction/stylesheets/index.styl
@@ -1,6 +1,6 @@
 @require '../../../components/clock'
 @require '../../../components/add_to_calendar'
-@require '../../../components/react/auction_block'
+@require '../../../components/react/AuctionBlock'
 @require '../../../components/my_active_bids'
 @require '../../../components/bid_status'
 @require '../../../components/artwork_item/stylesheets'
@@ -10,28 +10,28 @@
 @require '../../../components/stylus_lib'
 @require '../../../components/commercial_filter/stylesheets'
 
-@require '../components/artwork_browser/main/artwork/BidStatus'
-@require '../components/artwork_browser/main/artwork/GridArtwork'
-@require '../components/artwork_browser/main/artwork/ListArtwork'
-@require '../components/artwork_browser/main/artwork/MasonryArtwork'
-@require '../components/artwork_browser/main/ArtworkDisplay'
-@require '../components/artwork_browser/main/LoadingSpinner'
-@require '../components/artwork_browser/sidebar/ArtistFilter'
-@require '../components/artwork_browser/sidebar/BasicCheckbox'
-@require '../components/artwork_browser/sidebar/MediumFilter'
-@require '../components/artwork_browser/sidebar/RangeSlider'
-@require '../components/artwork_browser/header/HeaderDesktop'
-@require '../components/artwork_browser/header/HeaderMobile'
-@require '../components/artwork_browser/ArtworkBrowser'
+@require '../components/ArtworkBrowser/Main/Artwork/BidStatus'
+@require '../components/ArtworkBrowser/Main/Artwork/GridArtwork'
+@require '../components/ArtworkBrowser/Main/Artwork/ListArtwork'
+@require '../components/ArtworkBrowser/Main/Artwork/MasonryArtwork'
+@require '../components/ArtworkBrowser/Main/ArtworkDisplay'
+@require '../components/ArtworkBrowser/Main/LoadingSpinner'
+@require '../components/ArtworkBrowser/Sidebar/ArtistFilter'
+@require '../components/ArtworkBrowser/Sidebar/BasicCheckbox'
+@require '../components/ArtworkBrowser/Sidebar/MediumFilter'
+@require '../components/ArtworkBrowser/Sidebar/RangeSlider'
+@require '../components/ArtworkBrowser/Header/HeaderDesktop'
+@require '../components/ArtworkBrowser/Header/HeaderMobile'
+@require '../components/ArtworkBrowser/ArtworkBrowser'
 
-@require '../components/artwork_rail/ArtworkRail'
+@require '../components/ArtworkRail/ArtworkRail'
 
-@require '../components/layout/auction_info/AuctionInfo'
-@require '../components/layout/auction_info/Registration'
-@require '../components/layout/Banner'
-@require '../components/layout/Footer'
-@require '../components/layout/active_bids/MyActiveBids'
-@require '../components/layout/active_bids/ActiveBidItem'
+@require '../components/Layout/AuctionInfo/AuctionInfo'
+@require '../components/Layout/AuctionInfo/Registration'
+@require '../components/Layout/Banner'
+@require '../components/Layout/Footer'
+@require '../components/Layout/ActiveBids/MyActiveBids'
+@require '../components/Layout/ActiveBids/ActiveBidItem'
 
 .auction-body
   @media (max-width: responsive-mobile-width)
diff --git a/desktop/apps/auctions2/components/App.js b/desktop/apps/auctions2/components/App.js
index a6417901f..f37e20eae 100644
--- a/desktop/apps/auctions2/components/App.js
+++ b/desktop/apps/auctions2/components/App.js
@@ -2,7 +2,7 @@ import DOM from 'desktop/apps/auctions2/components/DOM'
 import Layout from 'desktop/apps/auctions2/components/Layout'
 import PropTypes from 'prop-types'
 import React, { Component } from 'react'
-import ResponsiveWindow from 'desktop/components/react/responsive_window'
+import ResponsiveWindow from 'desktop/components/react/ResponsiveWindow'
 import { Provider } from 'react-redux'
 
 export default class App extends Component {
diff --git a/desktop/apps/auctions2/reducers/index.js b/desktop/apps/auctions2/reducers/index.js
index 60d5fd927..95fea0f91 100644
--- a/desktop/apps/auctions2/reducers/index.js
+++ b/desktop/apps/auctions2/reducers/index.js
@@ -1,6 +1,6 @@
 import app, { initialState as appInitialState } from 'desktop/apps/auctions2/reducers/appReducer'
 import { combineReducers } from 'redux'
-import { composeReducers, responsiveWindowReducer } from 'desktop/components/react/responsive_window'
+import { composeReducers, responsiveWindowReducer } from 'desktop/components/react/ResponsiveWindow'
 
 export const initialState = {
   ...appInitialState
diff --git a/desktop/apps/consign/client/reducers.js b/desktop/apps/consign/client/reducers.js
index efe80b368..cf6a19a7a 100644
--- a/desktop/apps/consign/client/reducers.js
+++ b/desktop/apps/consign/client/reducers.js
@@ -5,7 +5,7 @@ import { composeReducers } from '../../../components/react/utils/composeReducers
 import { data as sd } from 'sharify'
 import { contains } from 'underscore'
 import { reducer as formReducer } from 'redux-form'
-import { responsiveWindowReducer } from '../../../components/react/responsive_window'
+import { responsiveWindowReducer } from '../../../components/react/ResponsiveWindow'
 import { routerReducer } from 'react-router-redux'
 
 const createAccountStep = 'createAccount'
diff --git a/desktop/apps/consign/client/submission.js b/desktop/apps/consign/client/submission.js
index 73a89916f..66a6b433b 100644
--- a/desktop/apps/consign/client/submission.js
+++ b/desktop/apps/consign/client/submission.js
@@ -1,5 +1,5 @@
 import analyticsMiddleware from './analytics_middleware'
-import ResponsiveWindow from '../../../components/react/responsive_window'
+import ResponsiveWindow from '../../../components/react/ResponsiveWindow'
 import React from 'react'
 import SubmissionFlow from '../components/submission_flow'
 import geo from '../../../components/geo/index.coffee'
diff --git a/desktop/apps/consign/test/components.js b/desktop/apps/consign/test/components.js
index 4e3aa5c84..904261b21 100644
--- a/desktop/apps/consign/test/components.js
+++ b/desktop/apps/consign/test/components.js
@@ -7,7 +7,7 @@ import UploadPhoto from '../components/upload_photo'
 import UploadedImage from '../components/uploaded_image'
 import React from 'react'
 import reducers from '../client/reducers'
-import { responsiveWindowAction } from '../../../components/react/responsive_window'
+import { responsiveWindowAction } from '../../../components/react/ResponsiveWindow'
 import { createStore } from 'redux'
 import { shallow } from 'enzyme'
 
diff --git a/desktop/components/react/auction_block/auction_block.js b/desktop/components/react/AuctionBlock/AuctionBlock.js
similarity index 96%
rename from desktop/components/react/auction_block/auction_block.js
rename to desktop/components/react/AuctionBlock/AuctionBlock.js
index edba58250..fe26aa507 100644
--- a/desktop/components/react/auction_block/auction_block.js
+++ b/desktop/components/react/AuctionBlock/AuctionBlock.js
@@ -1,6 +1,6 @@
 import PropTypes from 'prop-types'
 import React from 'react'
-import upcomingLabel from './utils/upcoming_label'
+import upcomingLabel from './utils/upcomingLabel'
 
 export default function AuctionBlock (props) {
   const {
diff --git a/desktop/components/react/auction_block/README.md b/desktop/components/react/AuctionBlock/README.md
similarity index 90%
rename from desktop/components/react/auction_block/README.md
rename to desktop/components/react/AuctionBlock/README.md
index 9243232bf..6c6d4bf59 100644
--- a/desktop/components/react/auction_block/README.md
+++ b/desktop/components/react/AuctionBlock/README.md
@@ -7,7 +7,7 @@ A React component that displays a single auction block. For use on its own (as i
 If using within a React context:
 
 ```javascript
-import { AuctionBlock } from 'components/react/auction_block'
+import { AuctionBlock } from 'components/react/AuctionBlock'
 
 class Page extends Component {
   render() {
@@ -29,7 +29,7 @@ If used within a non-React context, `index.jsx` will automatically mount provide
 ```
 
 ```coffeescript
-mountAuctionBlock = require 'components/react/auction_block'
+mountAuctionBlock = require 'components/react/AuctionBlock'
 
 class AuctionBlockContainer extends Backbone.View
   render: ->
diff --git a/desktop/components/react/auction_block/index.js b/desktop/components/react/AuctionBlock/index.js
similarity index 94%
rename from desktop/components/react/auction_block/index.js
rename to desktop/components/react/AuctionBlock/index.js
index 7b52aefd9..0c45c0637 100644
--- a/desktop/components/react/auction_block/index.js
+++ b/desktop/components/react/AuctionBlock/index.js
@@ -1,4 +1,4 @@
-import AuctionBlock from './auction_block'
+import AuctionBlock from './AuctionBlock'
 import invariant from 'invariant'
 import React from 'react'
 import ReactDOM from 'react-dom'
diff --git a/desktop/components/react/auction_block/index.styl b/desktop/components/react/AuctionBlock/index.styl
similarity index 100%
rename from desktop/components/react/auction_block/index.styl
rename to desktop/components/react/AuctionBlock/index.styl
diff --git a/desktop/components/react/auction_block/query.js b/desktop/components/react/AuctionBlock/query.js
similarity index 100%
rename from desktop/components/react/auction_block/query.js
rename to desktop/components/react/AuctionBlock/query.js
diff --git a/desktop/components/react/auction_block/utils/test/upcoming_label.js b/desktop/components/react/AuctionBlock/utils/test/upcomingLabel.test.js
similarity index 97%
rename from desktop/components/react/auction_block/utils/test/upcoming_label.js
rename to desktop/components/react/AuctionBlock/utils/test/upcomingLabel.test.js
index 739ce5ac2..dc4d4e6a1 100644
--- a/desktop/components/react/auction_block/utils/test/upcoming_label.js
+++ b/desktop/components/react/AuctionBlock/utils/test/upcomingLabel.test.js
@@ -1,4 +1,4 @@
-import upcomingLabel from '../upcoming_label'
+import upcomingLabel from '../upcomingLabel'
 
 describe('#upcomingLabel', () => {
   it('returns the correct label for an online-only auction that has yet to open', () => {
diff --git a/desktop/components/react/auction_block/utils/upcoming_label.js b/desktop/components/react/AuctionBlock/utils/upcomingLabel.js
similarity index 100%
rename from desktop/components/react/auction_block/utils/upcoming_label.js
rename to desktop/components/react/AuctionBlock/utils/upcomingLabel.js
diff --git a/desktop/components/react/current_auctions/current_auctions.js b/desktop/components/react/CurrentAuctions/CurrentAuctions.js
similarity index 95%
rename from desktop/components/react/current_auctions/current_auctions.js
rename to desktop/components/react/CurrentAuctions/CurrentAuctions.js
index 8fbe570b1..4a6cdd27e 100644
--- a/desktop/components/react/current_auctions/current_auctions.js
+++ b/desktop/components/react/CurrentAuctions/CurrentAuctions.js
@@ -1,7 +1,7 @@
 import PropTypes from 'prop-types'
 import React from 'react'
 import { partition, take } from 'underscore'
-import AuctionBlock from '../auction_block/auction_block'
+import AuctionBlock from '../AuctionBlock/AuctionBlock'
 
 export default function CurrentAuctions ({ auctionContextId, sales }) {
   const sortedSales = sortSales(sales, auctionContextId)
diff --git a/desktop/components/react/current_auctions/README.md b/desktop/components/react/CurrentAuctions/README.md
similarity index 89%
rename from desktop/components/react/current_auctions/README.md
rename to desktop/components/react/CurrentAuctions/README.md
index f4fcdb88e..cfadffd75 100644
--- a/desktop/components/react/current_auctions/README.md
+++ b/desktop/components/react/CurrentAuctions/README.md
@@ -7,7 +7,7 @@ A React component that displays a bar of four current Auctions.
 If using within a React context:
 
 ```javascript
-import { CurrentAuctions } from 'components/react/current_auctions'
+import { CurrentAuctions } from 'components/react/CurrentAuctions'
 
 class Page extends Component {
   render() {
@@ -28,7 +28,7 @@ If used within a non-React context, `index.jsx` will automatically mount provide
 ```
 
 ```coffeescript
-mountCurrentAuctions = require 'components/react/current_auctions'
+mountCurrentAuctions = require 'components/react/CurrentAuctions'
 
 class CurrentAuctionsContainer extends Backbone.View
   render: ->
diff --git a/desktop/components/react/current_auctions/index.js b/desktop/components/react/CurrentAuctions/index.js
similarity index 93%
rename from desktop/components/react/current_auctions/index.js
rename to desktop/components/react/CurrentAuctions/index.js
index 66edd0d12..d9c8cf5cf 100644
--- a/desktop/components/react/current_auctions/index.js
+++ b/desktop/components/react/CurrentAuctions/index.js
@@ -1,4 +1,4 @@
-import CurrentAuctions from './current_auctions'
+import CurrentAuctions from './CurrentAuctions'
 import invariant from 'invariant'
 import React from 'react'
 import ReactDOM from 'react-dom'
diff --git a/desktop/components/react/current_auctions/index.styl b/desktop/components/react/CurrentAuctions/index.styl
similarity index 100%
rename from desktop/components/react/current_auctions/index.styl
rename to desktop/components/react/CurrentAuctions/index.styl
diff --git a/desktop/components/react/current_auctions/query.js b/desktop/components/react/CurrentAuctions/query.js
similarity index 100%
rename from desktop/components/react/current_auctions/query.js
rename to desktop/components/react/CurrentAuctions/query.js
diff --git a/desktop/components/react/current_auctions/test/current_auctions_spec.js b/desktop/components/react/CurrentAuctions/test/CurrentAuctions.test.js
similarity index 96%
rename from desktop/components/react/current_auctions/test/current_auctions_spec.js
rename to desktop/components/react/CurrentAuctions/test/CurrentAuctions.test.js
index 491f8bfeb..882530091 100644
--- a/desktop/components/react/current_auctions/test/current_auctions_spec.js
+++ b/desktop/components/react/CurrentAuctions/test/CurrentAuctions.test.js
@@ -1,9 +1,9 @@
 import 'moment-timezone'
-import CurrentAuctions from '../current_auctions'
+import CurrentAuctions from '../CurrentAuctions'
 import React from 'react'
 import { render } from 'enzyme'
 
-describe('components/react/current_auctions_spec.js', () => {
+describe('components/react/CurrentAuctions.test.js', () => {
   const sales = [
     {
       id: 'climate-mobilization-benefit-auction-2017',
diff --git a/desktop/components/react/infinite_scroll/InfiniteScroll.js b/desktop/components/react/InfiniteScroll/InfiniteScroll.js
similarity index 100%
rename from desktop/components/react/infinite_scroll/InfiniteScroll.js
rename to desktop/components/react/InfiniteScroll/InfiniteScroll.js
diff --git a/desktop/components/react/masonry_grid/MasonryGrid.js b/desktop/components/react/MasonryGrid/MasonryGrid.js
similarity index 100%
rename from desktop/components/react/masonry_grid/MasonryGrid.js
rename to desktop/components/react/MasonryGrid/MasonryGrid.js
diff --git a/desktop/components/react/masonry_grid/Reveal.js b/desktop/components/react/MasonryGrid/Reveal.js
similarity index 100%
rename from desktop/components/react/masonry_grid/Reveal.js
rename to desktop/components/react/MasonryGrid/Reveal.js
diff --git a/desktop/components/react/masonry_grid/__tests__/Reveal.test.js b/desktop/components/react/MasonryGrid/__tests__/Reveal.test.js
similarity index 100%
rename from desktop/components/react/masonry_grid/__tests__/Reveal.test.js
rename to desktop/components/react/MasonryGrid/__tests__/Reveal.test.js
diff --git a/desktop/components/react/responsive_window/README.md b/desktop/components/react/ResponsiveWindow/README.md
similarity index 88%
rename from desktop/components/react/responsive_window/README.md
rename to desktop/components/react/ResponsiveWindow/README.md
index 821ce0201..4d667f8b4 100644
--- a/desktop/components/react/responsive_window/README.md
+++ b/desktop/components/react/ResponsiveWindow/README.md
@@ -7,7 +7,7 @@ A Redux-oriented React component that listens for changes to the window size and
 ```javascript
 // reducers.js
 import { combineReducers } from 'redux'
-import { composeReducers, responsiveWindowReducer } from 'components/react/responsive_window'
+import { composeReducers, responsiveWindowReducer } from 'components/react/ResponsiveWindow'
 
 function myMainReducer(state, action) {
   ...
@@ -21,7 +21,7 @@ export default combineReducers({
 })
 
 // my_app.js
-import ResponsiveWindow from 'components/react/responsive_window'
+import ResponsiveWindow from 'components/react/ResponsiveWindow'
 
 export default function MyApp() {
   return (
@@ -46,7 +46,7 @@ export default function App({ isMobile }) {
 ## Usage without Redux
 
 ```javascript
-import ResponsiveWindow from 'components/react/responsive_window'
+import ResponsiveWindow from 'components/react/ResponsiveWindow'
 
 function handleWindowChange(innerWidth) {
   console.log(innerWidth)
diff --git a/desktop/components/react/responsive_window/index.js b/desktop/components/react/ResponsiveWindow/index.js
similarity index 100%
rename from desktop/components/react/responsive_window/index.js
rename to desktop/components/react/ResponsiveWindow/index.js
diff --git a/desktop/components/react/responsive_window/test/responsiveWindow.test.js b/desktop/components/react/ResponsiveWindow/test/ResponsiveWindow.test.js
similarity index 100%
rename from desktop/components/react/responsive_window/test/responsiveWindow.test.js
rename to desktop/components/react/ResponsiveWindow/test/ResponsiveWindow.test.js
