// artsy-backbone-mixins/lib/fetch.js
// Generated by CoffeeScript 1.7.1
(function() {
  var ARTSY_URL, Backbone, parse, _;

  _ = require('underscore');

  Backbone = require('backbone');

  parse = require("url").parse;

  ARTSY_URL = '';

  module.exports = function(a) {
    ARTSY_URL = a;
    return module.exports.methods;
  };

  module.exports.methods = {
    fetchUntilEnd: function(options) {
      var fetchPage, opts, page, _ref;
      if (options == null) {
        options = {};
      }
      page = ((_ref = options.data) != null ? _ref.page : void 0) - 1 || 0;
      opts = _.clone(options);
      fetchPage = (function(_this) {
        return function() {
          var _ref1;
          return _this.fetch(_.extend(opts, {
            data: _.extend((_ref1 = opts.data) != null ? _ref1 : {}, {
              page: page += 1
            }),
            remove: false,
            complete: function() {},
            success: function(col, res) {
              if (typeof options.each === "function") {
                options.each(col, res);
              }
              if (res.length === 0) {
                if (typeof options.success === "function") {
                  options.success(_this);
                }
                return typeof options.complete === "function" ? options.complete(_this) : void 0;
              } else {
                return fetchPage();
              }
            },
            error: function() {
              if (typeof options.error === "function") {
                options.error.apply(options, arguments);
              }
              return typeof options.complete === "function" ? options.complete() : void 0;
            }
          }));
        };
      })(this);
      return fetchPage();
    },
    fetchSetItemsByKey: function(key, options) {
      if (options == null) {
        options = {};
      }
      return new Backbone.Collection(null).fetch({
        url: "" + ARTSY_URL + "/api/v1/sets?key=" + key,
        cache: options.cache,
        success: (function(_this) {
          return function(sets) {
            if (!sets.length) {
              return options.success(_this);
            }
            return new Backbone.Collection(null).fetch({
              url: "" + ARTSY_URL + "/api/v1/set/" + (sets.first().get('id')) + "/items",
              cache: options.cache,
              success: function(col) {
                _this.reset(col.toJSON());
                return options.success(_this);
              },
              error: options.error
            });
          };
        })(this),
        error: options.error
      });
    }
  };

}).call(this);
