// artsy-backbone-mixins/lib/image.js
// Generated by CoffeeScript 1.6.3
(function() {
  var _;

  _ = require('underscore');

  module.exports = function(secureImagesUrl, imagesUrlPrefix) {
    if (imagesUrlPrefix == null) {
      imagesUrlPrefix = 'http://static%d.artsy.net';
    }
    return {
      defaultImageVersion: function() {
        return (this.get('image_versions') || this.get('versions'))[0];
      },
      missingImageUrl: function() {
        return "/images/missing_image.png";
      },
      hasImage: function(version) {
        return _.contains(this.get('image_versions'), version) || _.contains(this.get('versions'), version);
      },
      imageUrl: function(version) {
        if (version == null) {
          version = this.defaultImageVersion();
        }
        if (this.hasImage(version)) {
          return this.sslUrl(this.get('image_url').replace(':version', version));
        } else {
          return this.missingImageUrl();
        }
      },
      sslUrl: function(url) {
        if (!(secureImagesUrl && imagesUrlPrefix)) {
          return url;
        }
        if (this.imagesUrlPrefixRE == null) {
          this.imagesUrlPrefixRE = new RegExp(imagesUrlPrefix.replace('%d', '\\d'));
        }
        return url.replace(this.imagesUrlPrefixRE, secureImagesUrl);
      }
    };
  };

}).call(this);
