// gestures/index.js
var interact = require('interact-js'),
    EventEmitter = require('events').EventEmitter;

function GestureDetector(target){
    this.target = target || document;

    interact.on('start', this.target, this.__boundStart = this._start.bind(this));

    interact.on('drag', this.target, this.__boundStart = this._drag.bind(this));

    interact.on('end', this.target, this.__boundEnd = this._end.bind(this));

    this.gestures = [];
}
GestureDetector.prototype = Object.create(EventEmitter.prototype);
GestureDetector.prototype.constructor = GestureDetector;
GestureDetector.prototype._start = function(interaction) {
    interaction.points = [];
};
GestureDetector.prototype._drag = function(interaction) {
    interaction.points.push({
        x: interaction.pageX,
        y: interaction.pageY
    });
};
GestureDetector.prototype._end = function(interaction) {
    var detector = this;
    this.gestures.forEach(function(gesture){
        var gestureName = gesture.call(this, interaction.points);
        if(gestureName){
            detector.emit('gesture', {
                name: gestureName,
                points: interaction.points
            });
        }
    });
};
GestureDetector.destroy = function(){
    interact.off('start', this.target, this.__boundStart);

    interact.on('end', this.target, this.__boundEnd);
};

module.exports = GestureDetector;
